README.md
[![Build Status](https://secure.travis-ci.org/fluffybunnies/glados.png)](http://travis-ci.org/fluffybunnies/glados)

# glados
Monitor remote resource for changes
- Use the cli wrapper to send an email notification or POST to a url when the resource has changed
- Diffs are calculated using kpdecker's [diff](https://www.npmjs.com/package/diff)
- Will not start polling until client binds to `'change'`
- Use the CLI Wrapper to take action on `'change'`
	- POST to an endpoint, or
	- send email notification (supports Gmail and SES)


## App
```javascript
var glados = require('glados')

var watcher = glados('http://example.com/sensitive_data.html', 1000*60)
.on('change',function(diff){
	// do something with diff
})
.on('error',function(err){
	// we have some problem connecting with resource
})
```


## CLI Wrapper
See api notes below for opts
```bash
node ./bin/cli.js http://example.com/sensitive_data.html -n60000 \
-p http://myapi.example.com/notify_change \
-m volcomstoner2689@gmail.com
```


## Forever Wrapper
Make sure this thing never goes down. Wraps cli.js with sturdiness.<br />
Kill the process, stop the forever, reboot the instance. Baby will still be kickin.<br />
Most recently polled data is saved to tmp file in case process dies and reboots.
```
./bin/forever.sh http://example.com/sensitive_data.html -n60000 \
-p http://myapi.example.com/notify_change \
-m ohsosexybrit@gmail.com
```


##### Kill Forever
Shut down what you started with `forever.sh`
```
./bin/kill_forever.sh http://example.com/sensitive_data.html
# or kill processes that contain -s (softmatch):
./bin/kill_forever.sh -s sensitive_data.html
```


## App Api

##### glados( ttl [, pollInterval] )
returns a `watcher` that will monitor changes of `url`<br />
_string_ `url` - remote resource to watch<br />
_int_ `pollInterval` - interval in ms to check resource for changes. defaults to 1 minute

##### watcher.stop()
Stop watching. No further events will be fired after this is called.

##### watcher.on('change', function(diff, data){})
Fired when a change has been detected in remote resource<br />
_Array_ `diff` - an array of data comprising added, removed, and unchanged chunks<br />
_Buffer_ `data` - the new data, unadulterated

##### watcher.on('connection', function(data){})
Fired on response from initial poll<br />
_Buffer_ `data` - the response data

##### watcher.on('poll', function(){})
Fired immediately prior to each poll. The first `'poll'` event occurs before `'connection'`

##### watcher.on('error', function(err){})
Catch errors in the stream


## CLI Api

##### Arguments
Applies to both `cli.js` and `forever.sh`
- `[target]`
	- url of resource to watch
- `-n` or `--interval`
	- pollInterval
	- optional, defaults to 1 minute
- `-p` or `--post`
	- endpoint to post `diff` on change event
	- optional
	- supports multiple
- `-m` or `--email`
	- email address to send `diff` on change event
	- optional
	- supports multiple


## Email
Copy `config.json` to `config.local.json` and add credentials

##### Gmail
Set `mailTransport` to 'gmail'<br />
Use an application password (not your personal account password)

##### SES
Set `mailTransport` to 'ses'<br />
Set your AWS secret and key, along with an approved From address


## To Do
- Using the softmatch option (`kill_forever.sh -s "substring"`) currently kills only the first match found, but deletes all matching saved data
	- Update script to kill all matches as advertised in readme
- Add screenshot of pretty email to readme


<!--
cd /root
git clone https://github.com/fluffybunnies/glados.git
cd glados
npm install
echo '' > ./config.local.json && vim ./config.local.json

node /root/glados/bin/cli.js http://www.huffingtonpost.com -n5000 \
-m volcomstoner2689@gmail.com --save

node /root/glados/bin/cli.js http://www.huffingtonpost.com -n5000 \
-p http://ace.fabfitfun.com/glados.php

/bin/bash /root/glados/bin/forever.sh http://data.iana.org/TLD/tlds-alpha-by-domain.txt \
-m volcomstoner2689@gmail.com

To Do
	- New sire module with simple index.php that generates content based on domain string
	- On TLD change, nab as many BrandName.NewTLDs as possible and point DNS to this server
		- Option to automate, but default to sending an email to self
		- If I click on confirm button, it'll start buying
		- If I don't click on confirm button within X time and the # new domains is reasonable, start buying
-->




CHANGELOG.md
## 1.0.0 (2015-07-09)

Features

	- Added a softmatch `-s` option for `kill_forever.sh`
	- Removed clutter in /var/log/glados.log
	- Bumped to 1.0.0: Features vetted, nothing new planned for near future

Bugs

	- `kill_forever.sh` was using arguments to generate a too-specific key, which could lead to forever not stopping



## 0.0.11 (2015-07-05)

Features

	- Added raw `data` argument to `'change'` event

Bugs

	- `data` passed to `'connection'` event is now a Buffer as advertised in docs



## 0.0.10 (2015-07-05)

Features

	- Save data locally when using `forever.sh` so we don't miss any changes while rebooting



## 0.0.7 (2015-07-05)

Features

	- Improved readme



## 0.0.6 (2015-07-05)

Features

	- Script to shutdown what `forever.sh` started: `kill_forever.sh`
	- Key forevers to target url instead of entire command. This allows argument adjustments without duplicating processes



## 0.0.5 (2015-07-05)

Features

	- Send plaintext diff as attachment in emails
	- Show only changed lines in email body
	- Set content-type header for `cli.js` POST handlers so php's `$_POST` parses the data automatically
	- + changelog



## 0.0.4 (2015-07-04)

Bugs

	- Escape html characters for html-compatible clients



## 0.0.1 (2015-07-04)

Features

	- Forever wrapper script


test/cli.js

var n = 0
,data1 = 'a\nb\nc\nd\ne\n'
,data2 = 'a\nb\nx\nd\ne\n'

require('http').createServer(function(req,res){
	n = !n;
	res.end(n ? data1 : data2);
}).listen(3001)


test/diff.js
/*
node ./test/diff.js
*/

var diff = require('diff')
,fs = require('fs')

var a = fs.readFileSync(__dirname+'/a.txt').toString();
var b = fs.readFileSync(__dirname+'/b.txt').toString();

console.log(
	require('../lib/cli_diff.js')(diff.diffLines(a,b))
)



test/test.js
var test = require('tape')
,glados = require('../')
,http = require('http')
,serverData = ''


test('worky',function(t){
	t.plan(5);

	var gotChange = false
	,gotPoll = false
	,s,failback

	s = startServer(function(){

		serverData = 'a\nb\nc\nd\ne';
		var watcher = glados('http://localhost:3001', 50)
		.on('change',function(){
			if (gotChange) return;
			gotChange = true;
			t.ok(true, 'got first change event');
		})
		.on('connection',function(data){ // connection event on first response
			t.ok(data instanceof Buffer, 'data passed to \'connection\' is a Buffer')
			t.ok(data.toString() == serverData, 'connection passeth data')
			serverData = 'a\nb\nx\nd\ne';
			watcher.on('poll',function(){ // poll event sent right before request
				if (gotPoll) return;
				gotPoll = true;
				watcher.on('change',function(diff, data){
					t.ok(diff[1].removed && diff[2].added, 'diff confirmed');
					t.ok(data instanceof Buffer, 'data passed to \'change\' is a Buffer')
					watcher.stop();
					s.close();
					clearTimeout(failback);
				})
			})
		})
		.on('error',function(err){
			t.fail(err)
		})

	})

	failback = setTimeout(function(){
		//s._connections && s.close();
		s.close();
		t.fail('timeout');
	},2000);

});

test('no change',function(t){
	t.plan(1);

	var s,watcher;
	s = startServer(function(){
		serverData = 'abcd\nefgh\n';
		watcher = glados('http://localhost:3001', 50)
		.on('change',function(){
			watcher.stop()
			s.close()
			t.fail('should not have triggered change event')
		})
	})

	setTimeout(function(){
		watcher.stop()
		s.close()
		t.ok(true, 'nothing changed')
	},500)

})

test('http error',function(t){
	t.plan(1);

	var s,failback;

	s = startServer(function(){
		serverData = 'a\nb\nc\nd\ne';
		var watcher = glados('http://localhost:3001', 50)
		.on('change',function(){
			t.fail('should have stopped (change1)')
		})
		.on('connection',function(){
			s.close()
			serverData = 'a\nb\nx\nd\ne';
			watcher.on('poll',function(){
				watcher.on('change',function(){
					t.fail('should have stopped (change2)')
				})
			})
		})
		.on('error',function(err){
			watcher.stop()
			clearTimeout(failback)
			t.ok(true, 'http error handled: '+err)
		})
	})

	failback = setTimeout(function(){
		s.close();
		t.fail('timeout');
	},2000);

})

function startServer(cb){
	return http.createServer(function(req,res){
		process.nextTick(function(){
			res.end(serverData);
		});
	}).listen(3001,cb);
}


lib/util.sh

if [ "`which realpath`" == "" ]; then
	realpath() {
		if [ ! -f "$1" ] && [ ! -d "$1" ]; then
			>&2 echo 'path does not exist'
		else
			dir=$1
			if [ -f "$1" ]; then
				dir=`dirname "$1"`
			fi
			path=`cd "${dir}";pwd`
			if [ -f "$1" ]; then
				path=$path/`basename "$1"`
			fi
			echo $path
		fi
	}
fi

if [ "$1" ]; then
	export __dirname=`dirname "$0"`
	export APP_PATH=`realpath "$__dirname/$1"`
	cd "$APP_PATH"
fi

crontab_add(){
  search=$1
  line=$2
  if [ ! "$line" ]; then line=$search; fi
  tmp=`mktemp`
  crontab -l | grep -v "$search" > $tmp
  echo "$line" >> $tmp
  crontab < $tmp 
  rm $tmp
}

crontab_remove(){
  search=$1
  tmp=`mktemp`
  crontab -l | grep -v "$search" > $tmp
  crontab < $tmp
  rm $tmp
}

locate_forever(){
	forevs=`which forever`
	if [ ! "$forevs" ]; then
		if [ -f /usr/local/bin/forever ]; then forevs=/usr/local/bin/forever
		elif [ -f /usr/bin/forever ]; then forevs=/usr/bin/forever
		elif [ -f /usr/bin/apt-get ]; then
			/usr/bin/apt-get install --assume-yes forever
			forevs=`which forever`
		fi
	fi
	echo $forevs
}


lib/send_email.js

var nodeMailer = require('nodemailer')
,config = require('../config.js')

module.exports = function(to, subject, body, cb){
	var transport,msg,from;
	if (config.mailTransport == 'ses') {
		transport = require('nodemailer-ses-transport')({
			accessKeyId: config.awsAccessKey
			,secretAccessKey: config.awsAccessSecret
			,region: config.awsRegion
		});
		from = config.sesMailFrom;
	} else if (config.mailTransport == 'gmail') {
		transport = {
			service: 'Gmail'
			,auth: {
				user: config.gmailUser
				,pass: config.gmailPass
			}
		};
	}
	if (!transport)
		return process.nextTick(function(){
			cb('missing mailTransport config');
		});

	if (typeof(to) == 'object') {
		msg = to;
		cb = subject;
	} else {
		msg = {
	    to: to
	    ,subject: subject
	    ,text: body
		};
	}
	if (!msg.from && from)
		msg.from = from;
	nodeMailer.createTransport(transport).sendMail(msg,cb);
}



index.js

var events = require('events')
,http = require('http')
,diff = require('diff')
,httpGet = require('./lib/http_get.js')
,defaultPollInterval = 1000*60

module.exports = function(url, pollInterval, useSavedData){
	if (isNaN(pollInterval = +pollInterval))
		pollInterval = defaultPollInterval;
	var w = new events.EventEmitter
		,someonesListening,connected
		,timeout,stopped

	// dont start watching until someone cares
	var on_ = w.on;
	w.on = function(evt){
		if (!someonesListening && evt == 'change') {
			someonesListening = true;
			process.nextTick(poll);
		}
		return on_.apply(w,arguments);
	}

	w.stop = function(){
		stopped = true;
		clearTimeout(timeout);
	}

	var lastPollData, undef;
	if (useSavedData) {
		lastPollData = useSavedData.toString();
		connected = true;
	}

	function poll(){
		w.emit('poll');
		httpGet(url,function(err, data){
			if (stopped) return;
			if (err) return w.emit('error',err);
			var thisPollData = data.toString();
			w.emit('_polled',data); if (stopped) return;
			if (!connected) {
				connected = true;
				w.emit('connection', data);
				if (stopped) return;
			}
			if (lastPollData != thisPollData && lastPollData !== undef) {
				var d = diff.diffLines(lastPollData,thisPollData);
				if (!d[1])
					return w.emit('error','new data but diff algo failed to notice');
				w.emit('change', d, data);
				if (stopped) return;
			}
			lastPollData = thisPollData;
			timeout = setTimeout(poll,pollInterval);
		})
	}

	return w;
}


